[mlir][VectorOps] Add `vector.interleave` operation

MacDue opened the pull request:
The interleave operation constructs a new vector by interleaving the elements from the trailing (or final) dimension of two input vectors, returning a new vector where the trailing dimension is twice the size.

Note that for the n-D case this differs from the interleaving possible with `vector.shuffle`, which would only operate on the leading dimension.

Another key difference is this operation supports scalable vectors, though currently a general LLVM lowering is limited to the case where only the trailing dimension is scalable.

Example:
```mlir
%0 = vector.interleave %a, %b
            : vector<[4]xi32>     ; yields vector<[8]xi32>
%1 = vector.interleave %c, %d
            : vector<8xi8>        ; yields vector<16xi8>
%2 = vector.interleave %e, %f
            : vector<f16>         ; yields vector<2xf16>
%3 = vector.interleave %g, %h
            : vector<2x4x[2]xf64> ; yields vector<2x4x[4]xf64>
%4 = vector.interleave %i, %j
            : vector<6x3xf32>     ; yields vector<6x6xf32>
```

MacDue committed:
[mlir][VectorOps] Add `vector.interleave` operation

The interleave operation constructs a new vector by interleaving the
elements from the trailing (or final) dimension of two input vectors,
returning a new vector where the trailing dimension is twice the size.

Note that for the n-D case this differs from the interleaving possible
with `vector.shuffle`, which would only operate on the leading dimension.

Another key difference is this operation supports scalable vectors,
though currently a general LLVM lowering is limited to the case where
only the trailing dimension is scalable.

Example:
```mlir
%0 = vector.interleave %a, %b
            : vector<[4]xi32>     ; yields vector<[8]xi32>
%1 = vector.interleave %c, %d
            : vector<8xi8>        ; yields vector<16xi8>
%2 = vector.interleave %e, %f
            : vector<f16>         ; yields vector<2xf16>
%3 = vector.interleave %g, %h
            : vector<2x4x[2]xf64> ; yields vector<2x4x[4]xf64>
%4 = vector.interleave %i, %j
            : vector<6x3xf32>     ; yields vector<6x6xf32>
```

dcaballe commented:
Thanks! It looks great!

> %0 = vector.interleave %a, %b
>            : vector<[4]xi32>     ; yields vector<[8]xi32>

I was thinking that we could return two vector results, one for the high and one for the low part of the interleave. However, I think keeping them together makes more sense at this level of abstraction.

dcaballe commented on mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:
```diff
@@ -478,6 +478,71 @@ def Vector_ShuffleOp :
+    returning a new vector where the trailing dimension is twice the size.
+
+    Note that for the n-D case this differs from the interleaving possible with
+    `vector.shuffle`, which would only operate on the leading dimension.
```
I thought we didn't even support n-D cases in `vector.shuffle`

MacDue replied:
It does (as can be seen in the examples at: https://mlir.llvm.org/docs/Dialects/Vector/#vectorshuffle-vectorshuffleop), it just does not do what you'd want :pensive:

dcaballe commented on mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp:
```diff
@@ -1734,6 +1734,70 @@ struct VectorSplatNdOpLowering : public ConvertOpToLLVMPattern<SplatOp> {
   }
 };
 
+struct VectorInterleaveOpLowering
```
doc

dcaballe commented on mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp:
```diff
@@ -1734,6 +1734,70 @@ struct VectorSplatNdOpLowering : public ConvertOpToLLVMPattern<SplatOp> {
+    }
+
+    // It's not possible to unroll a scalable dimension.
+    if (resultType.getScalableDims().front())
```
is this check correct? If so, could we make it easier to understand?

MacDue replied:
This is correct :)
This just looks at the scalable flag for the front dimension (which is the dimension that is unrolled below), and checks if it is true (i.e. scalable). If it is scalable it can't be unrolled.

dcaballe replied:
It would help if we add a comment here saying that we are unrolling only the leading dimension because this pattern is applied recursively. Even better, we could unroll all the dimensions at once to avoid recursion and applying partial unrolling when the rewrite may eventually fail. We could separate the base case in one pattern and the unrolling one in another pattern if that makes it clearer. We follow that approach often.

MacDue replied:
I've split out the unrolling to `LowerVectorInterleave`. Note: A incomplete unrolling is _not_ a bug, like if it was vector `vector.interleave %a, %b : vector<2x[4]x[4]xf32>` and it unrolled to `vector.interleave %a, %b : vector<[4]x[4]xf32>`, the general LLVM lowering can't handle that, but maybe something like ArmSME could then lower that to something architecture specific.

dcaballe commented on mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp:
```diff
@@ -1734,6 +1734,70 @@ struct VectorSplatNdOpLowering : public ConvertOpToLLVMPattern<SplatOp> {
+    auto loc = interleaveOp.getLoc();
+    Value result = rewriter.create<arith::ConstantOp>(
+        loc, resultType, rewriter.getZeroAttr(resultType));
+    for (int d = 0; d < resultType.getDimSize(0); d++) {
```
Per coding guidelines:
 - move ub to variable
 - use post-increment

dcaballe commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -6308,6 +6308,48 @@ bool WarpExecuteOnLane0Op::areTypesCompatible(Type lhs, Type rhs) {
+// The rank 1 case of vector.interleave on fixed-size vectors is equivalent to a
+// vector.shuffle, which (as an older op) is more likely to be matched by
+// existing pipelines.
+struct FoldRank1FixedSizeInterleaveOp : public OpRewritePattern<InterleaveOp> {
```
I think there is a point in using the interleave op as the canonical way to represent this shuffle pattern. It provides more information and it will keep it consistent with scalable.

dcaballe replied:
It should be ok to change an existing canonical representation if the new one is more explicit for a particular case.

> I don't think vector.interleave provides any more information in the case either, once you've matched the shuffle mask

Well, that's exactly the point! Some of the ops we have at this level are intended to simplify the analysis part of well-defined cases that are commonly used. Analyzing the shuffle mask is doable but it's adding complexity and we would have to exclude the non-interleave cases from the interleave ones when we have rewrites only targeting the interleave ones. 

> I would not be surprised if adding a vector.shuffle -> vector.interleave canonicalization broke something somewhere (maybe in a downstream project).

It's possible but `vector.shuffle` is used very late in the pipeline and we don't have so many rewrites targeting it. I think the benefit of having scalable and non-scalable aligned and more explicit representation for vector interleave patterns is worth the risk.

MacDue replied:
Sure, that makes sense, I'll try adding the inverse fold and see what breaks :)

dcaballe commented on mlir/test/Dialect/Vector/canonicalize.mlir:
```diff
@@ -2567,3 +2567,26 @@ func.func @load_store_forwarding_rank_mismatch(%v0: vector<4x1x1xf32>, %arg0: te
+func.func @fold_rank_1_vector_interleave(%arg0: vector<6xi32>, %arg1: vector<6xi32>) -> vector<12xi32> {
+  // CHECK: %[[ZIP:.*]] = vector.shuffle %[[LHS]], %[[RHS]] [0, 6, 1, 7, 2, 8, 3, 9, 4, 10, 5, 11] : vector<6xi32>, vector<6xi32>
+  // CHECK: return %[[ZIP]] : vector<12xi32>
+  %0 = vector.interleave %arg0, %arg1 : vector<6xi32>
```
Perhaps we should print the result type somehow? I know it can be inferred but some people prefer to make it easier to read and this op is not particularly verbose. WDYT?

c-rhodes replied:
> Perhaps we should print the result type somehow? I know it can be inferred but some people prefer to make it easier to read and this op is not particularly verbose. WDYT?

I think inference is nice from a builder perspective where building an op requires minimal info, but I do prefer if types are explicitly spelled out in the IR.

MacDue replied:
I personally find showing all the types here very repetitive, and less consistent with `vector.shuffle` which also infers the result.

dcaballe reviewed (commented):
It looks great! A few comments

c-rhodes commented on mlir/test/Dialect/Vector/canonicalize.mlir:
```diff
@@ -2567,3 +2567,26 @@ func.func @load_store_forwarding_rank_mismatch(%v0: vector<4x1x1xf32>, %arg0: te
+  return %0 : vector<12xi32>
+}
+
+// -----
+
+// CHECK-LABEL: func.func @fold_rank_0_vector_interleave(
+//  CHECK-SAME:     %[[LHS:.*]]: vector<f64>, %[[RHS:.*]]: vector<f64>)
+func.func @fold_rank_0_vector_interleave(%arg0: vector<f64>, %arg1: vector<f64>) -> vector<2xf64>
+{
+  // CHECK: %[[ZIP:.*]] = vector.shuffle %[[LHS]], %[[RHS]] [0, 1] : vector<f64>, vector<f64>
+  // CHECK: return %[[ZIP]] : vector<2xf64>
+  %0 = vector.interleave %arg0, %arg1 : vector<f64>
+  return %0 : vector<2xf64>
+}
```
took me a minute to figure out what's happening here, I'm not sure this test is really necessary. The fold relates to the rank of the result so this is no different to the test above.

MacDue replied:
Well, this is deliberately testing that the edge case of 0-D vectors works :)

c-rhodes commented on mlir/test/Dialect/Vector/canonicalize.mlir:
missing tests in: `mlir/test/Dialect/Vector/ops.mlir`

c-rhodes commented on mlir/test/Integration/Dialect/Vector/CPU/ArmSVE/test-scalable-interleave.mlir:
```diff
@@ -0,0 +1,25 @@
+// RUN: FileCheck %s
+
+func.func @entry() {
+  %f1 = arith.constant 1.0: f32
+  %f2 = arith.constant 2.0: f32
```
nit: inconsistent formatting, space after `:`
```suggestion
  %f1 = arith.constant 1.0 : f32
  %f2 = arith.constant 2.0 : f32
```

c-rhodes commented on mlir/test/Integration/Dialect/Vector/CPU/test-interleave.mlir:
```diff
@@ -0,0 +1,24 @@
+// RUN: FileCheck %s
+
+func.func @entry() {
+  %f1 = arith.constant 1.0: f32
+  %f2 = arith.constant 2.0: f32
```
nit: inconsistent formatting, space after `:`

MacDue committed:
Fixups

- Remove vector.interleave -> vector.shuffle canonicalization
- Add vector.shuffle -> vector.interleave canonicalization
- Split vector.interleave unrolling and LLVM lowering
  - Unrolling now done in LowerVectorInterleave.cpp
- Add missing tests to vector ops.mlir
- Fixed a few nits

hanhanW reviewed (commented):
I think this PR is not just adding `vector.interleave` operation. From my point of view, it has three changes:

1. Add `vector.interleave` op.
2. Add support for convert `vector.interleave` to llvm dialect.
3. Add support for unrolling `vector.interleave` op. 

Ideally we shoul split it into three PRs, which is easier for reviewers. It's okay for this one because it's already reviewed. Can you  add more context to the PR description?

banach-space commented:
> I think this PR is not just adding `vector.interleave` operation. From my point of view, it has three changes:
> 
> 1. Add `vector.interleave` op.
> 2. Add support for convert `vector.interleave` to llvm dialect.
> 3. Add support for unrolling `vector.interleave` op.
> 
> Ideally we shoul split it into three PRs, which is easier for reviewers. It's okay for this one because it's already reviewed. Can you add more context to the PR description?

@MacDue I would also find it easier to review if this was split :) (sorry, been away and smaller patches just make it easier to catch-up)

MacDue commented:
Would you both be okay with me just splitting this into multiple commits _within_ this PR? I find it becomes much more of a hassle then I've got 3+ branches and PRs to manage.

MacDue commented:
New PRs:
1. https://github.com/llvm/llvm-project/pull/80965
2. https://github.com/llvm/llvm-project/pull/80966
3. https://github.com/llvm/llvm-project/pull/80967
4. https://github.com/llvm/llvm-project/pull/80968
5. https://github.com/llvm/llvm-project/pull/80969

I've left the later PRs as drafts as it's easier to review/land them in order :)

dcaballe reviewed (commented):
I'm not sure if splitting the PR made the review simpler or harder given how bad github deals with stacked PRs :_(
